FLYonPIC v2.0

BrianMarre opened the pull request:
Second version of FLYonPIC
now includes:
- electronic collisional de-/excitation
- spontaneous deexcitation
- electronic collisional ionization
- autonomous ionization
- Stewart-Pyatt ionization potential depression(IPD)
- pressure ionization based on the IPD implementation

as well as a reworked atomic rate solver with adaptive time stepping

still missing:
- [x] FLYonPIC input specification
- [x] compile test
- [x] user facing description

requires
- [x] PR #4918 to be merged first
- [x] PR #4944 to be merged first

BrianMarre commented:
@PrometheusPi and @steindev for now no review required unless you really want to, a detailed examination together with @psychocoderHPC is already planned

psychocoderHPC commented on include/picongpu/particles/atomicPhysics2/kernel/SpawnIonizationMacroElectrons.kernel:
```diff
@@ -0,0 +1,366 @@
+            auto forEachFrameMaster = pmacc::lockstep::makeForEach<2u, T_Worker>(worker);
+            auto onlyMaster = pmacc::lockstep::makeMaster(worker);
+
+            IonFramePtr ionFrame = ionBox.getLastFrame(superCellIdx);
```
You must store this in shared memory if the master only is updating it to the next frame. Do not forget a sync after loading the last frame.
You can keep this in register if all worker update `ionFrame`, so changing

``` C++
onlyMaster(
                    [&ionFrame, &ionBox]()
                    {
                        ionFrame = ionBox.getPreviousFrame(ionFrame);
                        // no need to set "totalNumberMacroElectronsToSpawn = 0",
                        //  already guaranteed by break condition
                    });
```

to 

```C++
                        ionFrame = ionBox.getPreviousFrame(ionFrame);

```

is doing the same.

BrianMarre replied:
I see where I went wrong, thanks that would have taken me ages

psychocoderHPC commented on include/picongpu/particles/atomicPhysics2/rateCalculation/BoundBoundTransitionRates.hpp:
```diff
@@ -122,8 +121,8 @@ namespace picongpu::particles::atomicPhysics2::rateCalculation
                 g = A * math::log(U) + B + C / (U + a) + D / ((U + a) * (U + a)); // unitless
```
If you move `math::log(U)` before the loop the code will be faster `auto const logU = math::log(U);`

psychocoderHPC replied:
**Attention** possible division by zero in case where `(U + a) == 0` This should be handled!

psychocoderHPC replied:
**Attention2** `math::log(U)` can be NAN depending on the alpaka backend in cases where `U` is negative.
So we need to take care that `U` is guaranteed positive!

BrianMarre replied:
good catch, U is always >0 due to being the central value of a histogram bin of a histogram over the range [0, max], nevertheless will document the assumption

BrianMarre replied:
fixed

psychocoderHPC commented on include/picongpu/param/physicalConstants.param:
```diff
@@ -46,18 +46,22 @@ namespace picongpu
         // Electron properties
         /** unit: kg */
-        constexpr float_64 ELECTRON_MASS_SI = 9.109382e-31;
+        constexpr float_64 ELECTRON_MASS_SI = 9.1093837015e-31;
```
feel free to push the precision change in a separate PR before you atomic PR.

BrianMarre replied:
will do!

psychocoderHPC commented on include/picongpu/particles/atomicPhysics2/initElectrons/InitElectronAs.hpp:
```diff
@@ -0,0 +1,266 @@
+            float_X const phi = 2._X * static_cast<float_X>(picongpu::PI) * u;
+
+            float3_64 const directionVector
+                = float3_64(sinTheta * math::cos(phi), sinTheta * math::sin(phi), cosTheta);
```
`pmacc::math::sincos()` can be used here

BrianMarre replied:
done!

psychocoderHPC commented on include/picongpu/particles/atomicPhysics2/initElectrons/InitElectronAs.hpp:
```diff
@@ -0,0 +1,266 @@
+            //  unitless + (UNIT_MASS^2 * UNIT_SPEED^2)(not weighted)
+            //  /( UNIT_MASS^2 * UNIT_SPEED^2)(not weighted) = unitless
+            float_64 const normSquaredBetaIon_LabSystem
+                = 1._X / (1 + mSquaredCSquaredIon / momentumSquaredIon_LabSystem);
```
is this intended that `1 + ...` is used and `1` is integer?Our should it be `1.0_X`?

psychocoderHPC replied:
Attention ` 1._X` can be 32bit and the result is 64bit maybe `1.0` should be used everywhere.

BrianMarre replied:
You are right I intended 64 bit every where, oops

psychocoderHPC commented on include/picongpu/particles/atomicPhysics2/initElectrons/InitElectronAs.hpp:
```diff
@@ -0,0 +1,266 @@
+            auto beta = MatrixVector(betaIonSystem);
+
+            float_64 const gammaIonSystem
+                = math::sqrt(1. / (1. - pmacc::math::cPow(normSquaredBetaIon_LabSystem, 2u)));
```
**attention* possible division by zero if  `normSquaredBetaIon_LabSystem == 1.0`

BrianMarre replied:
fixed

BrianMarre replied:
from a maths and physics point of view an ion beta will always be less than zero, i.e. will never move with light speed.
For very highly relativistic ions floating point rounding errors may create beta values >= 1, but this are highly unlikelyin our simulations, since even a 1GeV hydrogen ion only reaches a beta of 0.7 .
Will create a comment but no check.

psychocoderHPC commented on include/picongpu/particles/atomicPhysics2/initElectrons/InitElectronAs.hpp:
```diff
@@ -0,0 +1,266 @@
+                // magnitude
+                math::sqrt(normSquaredBetaIon_LabSystem)
+                // direction
+                * static_cast<float3_64>(ion[momentum_] / pmacc::math::l2norm(ion[momentum_]));
```
**BUG??** division by zero in cases where the ion is not moving!!

BrianMarre replied:
good catch, fixed!

psychocoderHPC commented on include/picongpu/particles/atomicPhysics2/initElectrons/InitElectronAs.hpp:
```diff
@@ -0,0 +1,266 @@
+                for(uint32_t j = static_cast<uint32_t>(0u); j < static_cast<uint32_t>(3u); j++)
+                {
+                    lorentzMatrix.element(i, j) = (gammaIonSystem - 1.) * beta.element(i, static_cast<uint32_t>(0u))
+                        * beta.element(j, static_cast<uint32_t>(0u)) / normBetaSquared;
```
division by zero in cases where `normBetaSquared`

BrianMarre replied:
fixed

psychocoderHPC reviewed (changes requested):
I checked mostly for division by zero and found some critical parts, please check where it can happen and if function parameter must be `>0` to avoid a decision by zero please add it to the oxygen of the function that we do not forget about it.

psychocoderHPC commented on include/picongpu/particles/atomicPhysics2/atomicData/BoundFreeTransitionData.hpp:
```diff
@@ -17,7 +17,7 @@
  */
 
-#pragma once
+#pragma onc
```
@BrianMarre typo!

BrianMarre replied:
thanks, good catch!

psychocoderHPC commented on include/picongpu/particles/atomicPhysics2/initElectrons/Inelastic2BodyCollisionFromCoMoving.hpp:
```diff
@@ -0,0 +1,278 @@
+
+            // set to particle
+#pragma unroll
+            for(uint32_t i = 0u; i < picongpu::simDim; i++)
```
momentum has always 3 components! even in a 2D siulation

BrianMarre replied:
fixed

psychocoderHPC commented on include/picongpu/particles/atomicPhysics2/initElectrons/Inelastic2BodyCollisionFromCoMoving.hpp:
```diff
@@ -0,0 +1,278 @@
+
+                // set to particle
+#pragma unroll
+            for(uint32_t i = 0u; i < picongpu::simDim; i++)
```
should run over three components too

BrianMarre replied:
fixed

psychocoderHPC commented on include/picongpu/particles/atomicPhysics2/initElectrons/Inelastic2BodyCollisionFromCoMoving.hpp:
```diff
@@ -0,0 +1,278 @@
+            fillLorentzMatrix(beta, gammaIonSystem, normSquaredBetaIon_LabSystem, lorentzMatrix);
+
+            //      space components of Lorentz boost
+            MatrixVector momentumStarElectron_LabSystem;
```
This instance is not initialized because internal MAtrix is using an pmacc::Array which is not initializing its data

BrianMarre replied:
yes, which is why I am initializing it in the line below

psychocoderHPC commented on include/picongpu/particles/atomicPhysics2/interfaceDefintion.hpp:
```diff
@@ -0,0 +1,36 @@
+    // in-cell position to initialize macro particle to
+    CONST_VECTOR(
+        /*datatype*/ float_X,
+        /*dim*/ 3,
```
should be `simDim` or is atomic physics assuming always 3 components?

BrianMarre replied:
that is part of the rejected `one particle per superCell position functor` that will not go into the mainline anyway. but yes that should have probably accounted for simDim 2.

psychocoderHPC commented on include/picongpu/particles/atomicPhysics2/interfaceDefintion.hpp:
```diff
@@ -0,0 +1,36 @@
+        /*z*/ 0.0);
+
+    // cell in superCell to initialize macro particle to
+    CONST_VECTOR(uint32_t, 3, CellIdx, 0u, 0u, 0u);
```
should be `simDim` or is atomic physics assuming always 3 components?

BrianMarre replied:
fixed

BrianMarre replied:
will not go to mainline as it us part of a rejected pull request

psychocoderHPC commented on include/picongpu/particles/atomicPhysics2/initElectrons/Inelastic2BodyCollisionFromCoMoving.hpp:
```diff
@@ -120,60 +181,44 @@ namespace picongpu::particles::atomicPhysics2::initElectrons
 
-                electron[momentum_] = float3_X::create(0._X);
-                ion[momentum_] = float3_X::create(0._X);
+                electron[momentum_] = ion[momentum_];
```
@BrianMarre  you need to take the weighting into account because momentum in PIConGPU is weighted.

BrianMarre replied:
electron and ion weight are equal, due to the electron being created from the ion through ionization.

BrianMarre replied:
offline discussion; but psychocoderHPC nevertheless correct, while weights are equal, mass is not. I have to adjust the momentum for the different mass

BrianMarre commented:
@ikbuibui @chillenzer it is now at a state where a review from you would be helpful

chillenzer commented:
Wow! So, how many months do we get for 20k+ lines of code?

BrianMarre commented:
@psychocoderHPC do you have any idea why the CI fails at the KHI test/example? I am scratching my head on that one.

BrianMarre commented:
@psychocoderHPC please restart the KHI CI job, failed again with message "KILLED"

BrianMarre committed:
FLYonPICv2 with IPD and pressure ionization

BrianMarre committed:
documentation of atomicPhysics

BrianMarre committed:
add compile tests

BrianMarre commented:
@psychocoderHPC @chillenzer @ikbuibui @PrometheusPi @steindev ready for review

BrianMarre commented:
@psychocoderHPC @PrometheusPi CI fails in the KHI_growth rate test with,
``` bash
[ 78%] Built target picongpu-hostonly
[ 80%] Building CUDA object CMakeFiles/picongpu.dir/main.cpp.o
Killed
make[2]: *** [CMakeFiles/picongpu.dir/build.make:76: CMakeFiles/picongpu.dir/main.cpp.o] Error 137
make[1]: *** [CMakeFiles/Makefile2:243: CMakeFiles/picongpu.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```
I am not sure how to avoid this, I think psychocoderHPC suggested that this may be due to the CI running out of memory, but I am unsure why this would be specific to the FLYonPIC PR, other PRs CI tests pass fine and atomicPhysics should not be active at all in the KHI_growth test.

In summary my guess is that I am doing something wrong but I am unsure what exactly. Any help would be appreciated.

PrometheusPi commented:
If the single KHI_growthrate test job fails again, we have an issue with 64bit precision for that specific setup. 
This would require a detailed study, since this would then reveal unexpected side effects.

chillenzer commented:
Hi, here's a first batch of comments. These are more or less general comments, more detailed code comments as well as comments concerning the rest of the files will follow. In general, I will tag the comments, etc. with one of the following pre-fixes:

REQUEST: Necessary change (or at least necessary discussion).

REFACTOR: Nice-to-have but given the circumstances, we should use the strategy from [here](https://github.com/ComputationalRadiationPhysics/picongpu/issues/4912) and create an issue with such comments to address them later.

CONSCIOUS DECISION: A decision that has a trade-off. By commenting on this, I want to provoke a conscious decision that this is the best trade-off. You can resolve immediately (or with a short reasoning) if you think you thought this through.

QUESTION: More for my understanding than to entail code changes.

Please excuse that these are abbreviated to the edge of being impolite. They are copied more or less directly from my even shorter notes.

### Comments:

- CONSCIOUS DECISION: `atomicPhysics_Debug.param`
    - It surprised me that they exist and are different in the `tests` folder:
        - Some of the settings are backend-specific. Does this render the tests backend-specific?
        - Wouldn't we want to test the production version and not some debug version that might behave differently? (If only by different scheduling in multi-threading.)
        - The tests are compilation-only, so why activate debug there (assuming it's runtime debug)?
- REQUEST: Tests are missing `README`s.
- REFACTOR: An example would be nice:
	- Or are the tests also valid physical cases?
	- If so, having an example with more context and explanation would probably still be beneficial.
- CONSCIOUS DECISION: `atomicPhysics.param` is identical between default and `tests/compileAtomicPhysics`. Should we remove the latter?
- REQEUST: Is the FLYonPIC step diagram in the docs up to date? (At some point you showed me a version of this that you said was outdated, so just checking.)
	- Where does the arrow from "binElectrons" lead to?
	- What does this `KERNEL (/transitionType)` mean floating in the middle of nowhere?
	- What does the `true` node mean?
- REQUEST: Why are there any changes in `include/picongpu/fields/incidentField/profiles/DispersivePulse.hpp`?
- CONSCIOUS DECISION: Why remove DOI in `docs/source/usage/workflows/ionization.rst`?
- REQUEST: Unnecessary and questionable changes in `docs/source/models/collisional_ionization.rst`.

PrometheusPi reviewed (changes requested):
The CI test (`g++-11_cuda11.7_boost1.78.0_tests.KHI_growthRate`, KHI growth rate test cmake preset 1) fails consistently. There seems to be a consistent bug with a side dependency we do not yet understand. 
Until this side effect is not understood and solved, this pull request should be blocked.

BrianMarre commented:
@chillenzer
REQUEST FLYonPIC diagram:
- FLYonPIC diagram is updated
- good catch, I missed the extra floating text and too short life line for the electron histogram, fixed
- the true node indicates a `while(true)` , did not find a better way to draw it with PlantUML

REQUEST changes in ``docs/source/models/collisional_ionization.rst`` and ``include/picongpu/fields/incidentField/profiles/DispersivePulse.hpp``:

- I am not sure what exactly you mean, according to Github the PR does not touch ``docs/source/models/collisional_ionization.rst`` or ``include/picongpu/fields/incidentField/profiles/DispersivePulse.hpp``?

REFACTOR add Examples:
Well, the fundamental problem is that the atomic data input files I am using are not public, therefore we are unable to provide a setup that works at runtime.

BrianMarre committed:
fix diagram atomicPhysics step

BrianMarre committed:
update outdated comments in atomicPhysics_Debug.param

BrianMarre committed:
prevent compilation of atomicpHysics if not necessary v2

chillenzer commented on docs/source/models/atomic_physics.rst:
```diff
@@ -0,0 +1,230 @@
+
+This allows a more accurate prediction of ionization dynamics and the prediction of the atomic state distributions in non-thermal, transient plasmas, for example high-intensity short-pulse laser plasma interactions.
+
+FLYonPIC is base on the [FlyCHK]_ atomic model and currently implements the following processes:
```
```suggestion
FLYonPIC is based on the [FlyCHK]_ atomic model and currently implements the following processes:
```

BrianMarre replied:
fixed

chillenzer commented on docs/source/models/atomic_physics.rst:
```diff
@@ -0,0 +1,230 @@
+FLYonPIC is based on tracking the atomic state of ions as a particle attribute of ion macro particles, with each "macro-ion" representing a phase space sample of exactly one atomic state.
+
+.. note::
+   FLYonPIC tracks only FlyCHK super configurations.
+
+   i.e. FLYonPIC distinguishes only between states differing in the number of electron contained in atomic shells
```
```suggestion
   FLYonPIC tracks only FlyCHK super configurations, i.e., FLYonPIC distinguishes only between states differing in the number of electrons contained in atomic shells.
```

BrianMarre replied:
fixed

chillenzer commented on docs/source/models/atomic_physics.rst:
```diff
@@ -0,0 +1,230 @@
+
+Over the set of all transitions, atomic state dynamics may be represented as a matrix of rates describing the change of the entire atomic state population with time.
+
+This rate equation is solved in FLYonPIC by a monte-carlo solver for every macro-ion individually and time dependent.
```
```suggestion
This rate equation is solved in FLYonPIC by a Monte Carlo solver for every macro-ion individually and time dependent.
```

BrianMarre replied:
fixed

chillenzer commented on include/picongpu/particles/atomicPhysics/electronDistribution/LocalHistogramField.hpp:
```diff
@@ -0,0 +1,130 @@
+        using Histogram = T_Histogram;
+
+    private:
+        /// @todo should these be private?
```
QUESTION: What's the status of this?

BrianMarre replied:
I would keep them private, but psychocoder's opinion differed in the past. therefore marked for discussion

BrianMarre replied:
removed comment, will keep the string private to prevent name changes after creation

chillenzer commented on include/picongpu/param/speciesAttributes.param:
```diff
@@ -114,17 +115,201 @@ namespace picongpu
+     *  charge state!
      *
      * @todo connect default to proton number
+     * @todo change to uint8_t or uint32_t
```
QUESTION: What's the status of this?

BrianMarre replied:
this has been discussed in the #4228, last agreement was that yes we want to change boundElectrons to `uint`, but this should be done in a separate PR

chillenzer commented on include/picongpu/param/speciesAttributes.param:
```diff
@@ -114,17 +115,201 @@ namespace picongpu
+     *
+     * file gives parameter of atomic physics charge states for ion species
+     * - must follow PIConGPU specific formatting
+     *      @todo write specification, Brian Marre, 2023
```
QUESTION: What's the status of this?

BrianMarre replied:
done, will remove

chillenzer commented on include/picongpu/param/speciesAttributes.param:
```diff
@@ -114,17 +115,201 @@ namespace picongpu
+     * - must follow PIConGPU specific formatting
+     *      @todo write specification, Brian Marre, 2023
+     * - files are not shipped with PIConGPU, see documentation for creation and sources
+     *      @todo create documentation, Brian Marre, 2022
```
QUESTION: What's the status of this?

BrianMarre replied:
done, I removed the todo

chillenzer commented on include/picongpu/param/speciesAttributes.param:
```diff
@@ -114,17 +115,201 @@ namespace picongpu
+     *
+     * file gives parameter of atomic physics atomic states for ion species
+     * - must follow PIConGPU specific formatting
+     *      @todo write specification, Brian Marre, 2023
```
QUESTION: What's the status of this?

BrianMarre replied:
done removed the todo

chillenzer commented on include/picongpu/param/speciesAttributes.param:
```diff
@@ -114,17 +115,201 @@ namespace picongpu
+     * - must follow PIConGPU specific formatting
+     *      @todo write specification, Brian Marre, 2023
+     * - files are not shipped with PIConGPU, see documentation for creation and sources
+     *      @todo create documentation, Brian Marre, 2022
```
QUESTION: What's the status of this?

BrianMarre replied:
done, removed the todo

chillenzer commented on include/picongpu/param/speciesAttributes.param:
```diff
@@ -114,17 +115,201 @@ namespace picongpu
+     * - setting an atomic states the pressure ionization state equal to the state disables pressure ionization
+     *      for this states
+     * - file must follow PIConGPU specific formatting
+     *      @todo write specification, Brian Marre, 2023
```
Okay... this applies just for all todos: QUESTION: What's the status of this?

BrianMarre replied:
done, removed the todo

chillenzer replied:
Okay. Did you go through `grep -i todo -r .` and check the status of everything?

chillenzer commented on include/picongpu/particles/atomicPhysics/kernel/BinElectrons.kernel:
```diff
@@ -0,0 +1,104 @@
+         *
+         * @param worker object containing the device and block
+         *  information, passed by PMACC_KERNEL call
+         * @param areMapping mapping of blockIndex to block superCell index
```
```suggestion
         * @param areaMapping mapping of blockIndex to block superCell index
```

BrianMarre replied:
fixed

chillenzer commented on include/picongpu/simulation/stage/AtomicPhysics.hpp:
```diff
@@ -0,0 +1,604 @@
+
+                pmacc::DataConnector& dc = pmacc::Environment<>::get().DataConnector();
+
+                // TimeRemainingSuperCellField
+                auto& localTimeRemainingField = *dc.get<S_TimeRemainingField>("LocalTimeRemainingField");
+                DataSpace<picongpu::simDim> const fieldGridLayoutTimeRemaining
+                    = localTimeRemainingField.getGridLayout().sizeWithoutGuardND();
+
+                // FoundUnboundIonSuperCellField
+                auto& localFoundUnboundIonField = *dc.get<S_FoundUnboundField>("LocalFoundUnboundIonField");
+                DataSpace<picongpu::simDim> const fieldGridLayoutFoundUnbound
+                    = localFoundUnboundIonField.getGridLayout().sizeWithoutGuardND();
+
+                // ElectronHistogramOverSubscribedSuperCellField
+                auto& localElectronHistogramOverSubscribedField
+                    = *dc.get<S_OverSubscribedField>("LocalElectronHistogramOverSubscribedField");
+                DataSpace<picongpu::simDim> const fieldGridLayoutOverSubscription
+                    = localElectronHistogramOverSubscribedField.getGridLayout().sizeWithoutGuardND();
```
REFACTOR: For these variables it would have been nice to have a
```C++
struct LocalField
    field;
    layout;
    
    LocalField(&dc) : field{*dc.get...}, layout{field.getGridLayout()...}
};
```
to have these 14 lines down to three.

REQUEST: Also: Remove comments. They are redundant with the (quite good!) naming.

BrianMarre replied:
comments removed

chillenzer commented on include/picongpu/simulation/stage/AtomicPhysics.hpp:
```diff
@@ -0,0 +1,604 @@
+
+                uint16_t counterSubStep = 0u;
+                // atomicPhysics sub-stepping loop, ends when timeRemaining<=0._X
+                while(true)
```
REFACTOR: I would have preferred an actual condition or even better a for-loop here. Reason:
 1. It is hard to comprehend, check and verify the control flow with all those hidden `break`
 statements scattered around.
 2. In my head the following three have different semantics:

    - while(true):
      Infinite event loop that's supposed to run forever or until terminated by an external event.
    - while(condition):
      An iterative algorithm that does not progress but has a boolean success criterion, e.g.,
      hit-or-miss algorithms that return on the first hit.
    - for(...)
      An iterative algorithm that progresses in some meaningful way (not necessarily by `++i`)

 Admittedly, the border between while(condition) and for(...) is relatively vague and there is room
 for interpretation and personal preference there. But while(true) clearly communicates something
 different.

BrianMarre replied:
A for-loop is not possible, since we do not know in advance how many sub step we will do.

Fundamentally the entire loop body needs to be executed at least once therfore the best fit would be a do-while loop, but I am unsure how often people actually see do-whiles.

Open to changing it to do-while loops.

BrianMarre replied:
refactored to use do-while loops

chillenzer replied:
Offline discussion yielded some good results but just for the record: A `for` loop is exactly the same as a `while` loop except for the fact that you have the option to give it a bit more structure by using the initialisation and increment field.

```C++
for(<initialisation>, <condition>, <increment>) {
    ...
}
```
is identical to
```C++
<initialisation>
while(<condition>) {
    ...
    <increment>
}
```
(except for scoping).

And the other way round
```C++
while(<condition>) {
    ...
}
```
is the same as
```C++
for (; <condition>; ) {
    ...
}
```

The choice is really about what you want to communicate.

chillenzer commented on include/picongpu/simulation/stage/AtomicPhysics.hpp:
```diff
@@ -0,0 +1,604 @@
+                // atomicPhysics sub-stepping loop, ends when timeRemaining<=0._X
+                while(true)
+                {
+                    // debug only
```
REQUIRED: This does not look like "debug only". Remove comment or statement.

BrianMarre replied:
done

chillenzer commented on include/picongpu/traits/attribute/GetMass.hpp:
```diff
@@ -34,7 +34,8 @@ namespace picongpu
              * @param particle a reference to a particle
-             * @return mass of the makro particle
+             *
+             * @return mass of the macro particle, UNIT_MASS
```
QUESTION: Why that addition? Is this pertinent?

BrianMarre replied:
I changed the GetMass documentation to fit to the documentation style used for atomicPhyisc returns as it is used repeatedly in the atomcPhysics,  and while I was at it I fixed the typo.

Could be moved to separate PR, or removed, I am open to both.

BrianMarre replied:
In general the problem I ran into is that atomicPhysics needs a different normalization than the PIC-loop, since
atomicPhysics deals with fundamentally much smaller quantities and therefore I had to sometime use different units than the main code to avoid numerical errors.

Consequently I had to document what the unit the return values of some of my functions were and to make less errors in the internal unit conversions I created documentation to keep track of units.

psychocoderHPC replied:
@chillenzer not strongly necessary but okisch.

chillenzer commented on include/pmacc/mappings/kernel/MappingDescription.hpp:
```diff
@@ -87,6 +87,11 @@ namespace pmacc
 
+        HDINLINE DataSpace<DIM> getGridSuperCellsWithoutGuards() const
+        {
+            return this->gridSuperCells - 2 * this->guardingSuperCells;
```
QUESTION: Does the PMAcc model guarantee that the guards are symmetric?

BrianMarre replied:
I actually not sure, @psychocoderHPC care to chime in?

psychocoderHPC replied:
Guarding supercells can be different per dimension.

Why is this change here in this PR? looks like it is already merged.

BrianMarre replied:
@psychocoderHPC But are they the same for the two different directions of one dimension? 

I would assume yes, since MappingDescription stores only one value per dimension.

Note: This method is required by in the FLYonPIC SuperCell fields which are newly added with this PR, therefore this change also belongs to this PR

psychocoderHPC replied:
Yes guard size is per dimension and equal for the negative and positive direction.

chillenzer commented on share/picongpu/tests/compileAtomicPhysics/include/picongpu/param/atomicPhysics.param:
```diff
@@ -0,0 +1,101 @@
+    //}
+
+    // atomicPhysics rate solver settings
+    //{
```
REFACTOR: Do I understand this correctly that the curly braces are kind of creating an implicit subsection here? If so, why not use a struct and make it explicit? If not, this is likely deada code and should be removed.

BrianMarre replied:
they are supposed to indicate that the comment describes several lines, start and end indicated by the braces.
I copied this style from psychocoder but would be open to change it

BrianMarre replied:
followed your suggestion, and created a param struct

chillenzer commented on share/picongpu/tests/compileAtomicPhysics/include/picongpu/param/atomicPhysics.param:
```diff
@@ -0,0 +1,101 @@
+     * see picongpu/particles/atomicPhysics/atomicData/AtomicData.hpp for further information.
+     */
+    using AtomicData_Ar = particles::atomicPhysics::atomicData::AtomicData<
+        uint32_t, // T_Number
+        float_X, // T_Value
+        uint32_t, // T_CollectionIndex
+        ConfigNumber_Argon, // T_ConfigNumber, data type used for storage
+        float_64, // T_Multiplicities, data type used for storage
+        true, // T_electronicExcitation
+        true, // T_electronicDeexcitation
+        true, // T_spontaneousDeexcitation
+        true, // T_electronicIonization
+        true, // T_autonomousIonization
+        false>; // T_fieldIonization
```
REFACTOR: Could these be enums such that we say `ElectronicExcitation::On` and so on?

BrianMarre replied:
yes they could, but that entail changing a lot of kernels since they are used to toggle code branches on and off.

Would want to push that to a future refactor.

chillenzer commented on share/picongpu/tests/compileAtomicPhysics/include/picongpu/param/speciesInitialization.param:
```diff
@@ -0,0 +1,53 @@
+         *
+         * the functors are called in order (from first to last functor)
+         */
+        using InitPipeline = boost::mp11::mp_list<
```
QUESTION: Why not use a `pmacc::mp_list<>` here as per default? It's currently the same but it might not be in the future!

BrianMarre replied:
the input is based on an older version, did not notice that we changed this bit, good catch fixed

chillenzer commented on share/picongpu/tests/compileSparser/include/picongpu/param/speciesInitialization.param:
```diff
@@ -0,0 +1,53 @@
+         *
+         * the functors are called in order (from first to last functor)
+         */
+        using InitPipeline = boost::mp11::mp_list<
```
same here

BrianMarre replied:
fixed

chillenzer reviewed (changes requested):
First batch of comments. From what I've just typed here, I'd say pretty good work so far. Mostly typos and stuff... But there's more to come! (... probably, well, I haven't checked yet. That's the whole point, you see.)

chillenzer commented:
>  I am not sure what exactly you mean, according to Github the PR does not touch `docs/source/models/collisional_ionization.rst` or `include/picongpu/fields/incidentField/profiles/DispersivePulse.hpp`?

Okay. Then, ignore it. I pulled one specific commit last week Thursday night which I also found is outdated by now.

> REFACTOR add Examples: Well, the fundamental problem is that the atomic data input files I am using are not public, therefore we are unable to provide a setup that works at runtime.

Okay. (Well, not okay philosophically and concerning open science, etc., but that's probably not something we want to solve before merging this.)

BrianMarre committed:
remove solved todos

BrianMarre committed:
update param to use pmacc::mp_list

BrianMarre committed:
remove comments

BrianMarre committed:
prevent compiling kernel v3

BrianMarre committed:
remove debug console output

psychocoderHPC commented:
The failing CI test is the KHi with double precision. It is running into the time limit of 3h.

psychocoderHPC commented:
@BrianMarre please rebase against the dev branch. We removed ISAAC support, maybe this helps to pass the CI. This will reduce compile overhead, if not there is an issue with 64bit PIConGPU together with atomic physics.

BrianMarre committed:
fixing reviewer comments

BrianMarre committed:
switch to do-while loops

psychocoderHPC commented on include/picongpu/simulation/stage/AtomicPhysics.hpp:
```diff
@@ -350,63 +352,61 @@ namespace picongpu::simulation::stage
-                            if constexpr(picongpu::atomicPhysics::debug::kernel::rollForOverSubscription::
-                                             PRINT_DEBUG_TO_CONSOLE)
+                            // remove overSubscription
+                            bool noHistogramOverSubscribed = false;
```
no double negation please!

BrianMarre replied:
outdated, entire section has been refactored.

psychocoderHPC commented on include/picongpu/simulation/stage/AtomicPhysics.hpp:
```diff
@@ -468,16 +466,13 @@ namespace picongpu::simulation::stage
-                    {
-                        break;
-                    }
+                    timeRemainingLessThanZero = deviceLocalReduce(
```
no need to reflect the condition in the name. The name should be `stopExecuion`, `isValid...`

BrianMarre replied:
outdated, entire section has been refactored.

psychocoderHPC commented on include/picongpu/simulation/stage/AtomicPhysics.hpp:
```diff
@@ -302,8 +303,9 @@ namespace picongpu::simulation::stage
-                    // reject overSubscription loop, ends when no histogram bin oversubscribed
-                    while(true)
+                    // choose transition loop
+                    bool histogramOverSubscribed = true;
```
later you have a variable named `noHistogramOverSubscribed`. The namings are confusing, they sounds euqal but most likely they are not.

BrianMarre replied:
outdated, entire section has been refactored.

psychocoderHPC committed:
refactor atomic physics stage

Seperate control flow loop from sub stage calls.

Co-authored-by: BrianMarre <b.marre@hzdr.de>

psychocoderHPC commented on include/picongpu/particles/atomicPhysics/ionizationPotentialDepression/kernel/CalculateIPDInput.kernel:
```diff
@@ -166,4 +172,35 @@ namespace picongpu::particles::atomicPhysics::ionizationPotentialDepression::ker
     };
+
+    template<>
+    struct CalculateIPDInputKernel<0u>
```
please remove, it looks like it is not required

BrianMarre replied:
done

BrianMarre committed:
fixes

BrianMarre committed:
refactor atomic physics species configuration

BrianMarre committed:
update species filters with new species config

psychocoderHPC commented on include/picongpu/param/speciesAttributes.param:
```diff
@@ -147,146 +147,12 @@ namespace picongpu
-    alias(isAtomicPhysicsIon);
+    //! configuration of species for atomicPhysics algorithm, see picongpu/particles/atomicPhysics/ParticleType.hpp for
+    //! details
+    alias(atomicPhysics)
```
The CI says that `atomicPhysics` is colliding with the namespace `atomicPhysics`. Unfortunally you need to change one.

psychocoderHPC replied:
maybe change all namespaces to `atomicPhy`? This will be a simple search and replace over the code (2min work).

psychocoderHPC replied:
wait, maybe you missed the semicolon only should be `alias(atomicPhysics);`

BrianMarre replied:
well it was both, fixed

BrianMarre committed:
reviewer comments

BrianMarre committed:
update examples

psychocoderHPC committed:
introduce particle type traits

- add traits to get, filterBy particle type
- add particle type tags for easy meta template programming
- add collection for attributes needed by atomicPhysics ions
- rename alias `atomicPhysics_` to `atomicPhysicsParticle`

Co-authored-by: BrianMarre <b.marre@hzdr.de>

BrianMarre commented:
I checked that 94f1aab still fulfills the atomicPhysicsCI test

psychocoderHPC commented:
@BrianMarre please add (seperate PR) the svg or source data of your big plot you added as png. The png can not be editied.
